test: Implicitly sync after generate* to preempt races and intermittent test failures

maflcko opened the pull request:
The most frequent failure in functional tests are intermittent races. Fixing such bugs is cumbersome because it involves:
* Noticing the failure
* Fetching and reading the log to determine the test case that failed
* Adding a `self.sync_all()` where it was forgotten
* Spamming out a pr and waiting for review, which is already sparse

Also, writing a linter to catch those is not possible, nor is review effective in finding these bugs prior to merge.

Fix all future intermittent races caused by a missing sync_block call by calling `sync_all` implicitly after each `generate*`, unless opted out. This ensures that the code is race-free (with regards to blocks) when the tests pass once, instead of our current approach where the code can never be guaranteed to be race-free.

Commit:
test: Create TestNode._rpc_dispatch method

Commit:
test: Pass default_sync_fun to TestNode constructor

practicalswift commented:
Concept ACK: thanks for doing this - intermittent test failures are the worst!

Bike shed nit: `fn` or `func` might be a better (but less fun) abbreviation than `fun` :)

Commit:
test: Implicitly sync after generate*, unless opted out

Commit:
scripted-diff: Remove redundant sync_all

-BEGIN VERIFY SCRIPT-
perl -0777 -pi -e 's/(generate[^\n]*\)[^\n]*)(\n|\s)+self.sync_.*\(\)\n/\1\n/g' $(git grep -l generate ./test)
-END VERIFY SCRIPT-

promag reviewed (commented):
Concept ACK.

How about for now:
```
try_dispatch_with_sync
    try
        dispatch
    except
        sync_all
        dispatch
        log with print_stack
```

MarcoFalke commented on test/functional/test_framework/test_framework.py:
```diff
@@ -456,6 +456,7 @@ def get_bin_from_version(version, bin_name, bin_default):
         assert_equal(len(binary_cli), num_nodes)
         for i in range(num_nodes):
             test_node_i = TestNode(
+                self,
```
I didn't do that because `self.generate(1, 1)` looks unreadable. Enforcing named args makes everything verbose.

promag replied:
Agree.

promag replied:
See 8f8df56949e5e6558ff7b38bea51fee7acbcc129 where self.sync_all is passed to TestNode instead of self.

MarcoFalke commented on test/functional/test_framework/test_node.py:
```diff
@@ -296,9 +296,27 @@ def wait_for_cookie_credentials(self):
         self._raise_assertion_error("Unable to retrieve cookie credentials after {}s".format(self.rpc_timeout))
 
-    def generate(self, nblocks, maxtries=1000000):
+    def generateblock(self, *args, sync_fun=True, **kwargs):
```
That would make calling code tremendously verbose, no?

MarcoFalke replied:
Maybe I am misunderstanding. Do you have a working diff that I can steal?

amitiuttarwar commented:
Concept ACK! 

very nice, thank you

MarcoFalke commented:
rebased

MarcoFalke commented:
rebased

laanwj commented:
Concept ACK, rooting out a cause of intermittent issues is great.

MarcoFalke commented:
rebased

jonasschnelli commented:
Nice!
utACK faeef7a61307b99654a445dffe05b4aed561639d

MarcoFalke commented:
Rebased. Should be trivial to re-ACK

sipa commented on test/functional/test_framework/test_node.py:
```diff
@@ -62,14 +62,14 @@ class TestNode():
                 the node starts.
         """
 
+        self.framework = framework
```
It seems a bit ugly to have this bidirectional reference between the framework and the test nodes.
Would it be sufficient to instead pass and store a lambda "sync function"?

MarcoFalke replied:
Thanks, done

jonatack commented:
Code review ACK, needs rebase.

MarcoFalke commented:
Rebased. Should be trivial to re-ACK, as most conflicts were in the scripted diff.

MarcoFalke commented:
Rebased

MarcoFalke commented:
Rebased

jnewbery commented:
Concept ACK

jnewbery commented:
What do you think about adding a `generate()` method to `TestFramework` that takes the index of the node that you want to generate on? It seems strange to pass a TestFramework function to the constructor of the TestNode.

MarcoFalke commented:
That would require changing *all* calls to `generate*` and also disable `TestNode::generate*` to avoid races creeping in by accident. Happy to do that refactor if reviewers think it is worth it. I am mostly interested in the functional changes here.

jnewbery commented:
> Happy to do that refactor if reviewers think it is worth it.

I think it's worth it, to avoid adding an implicit dependency from TestNode onto TestFramework.

MarcoFalke commented:
Done in #22567. (Different pr, because the discussion here wasn't substantial, mostly GitHubs own spam, and I wanted to keep this version in case someone decides against #22567).
